File: Plugin/BlenderBridge/gmr/sidecar.js
'use strict';
/*
 * GMR ex 模块 - 推理 sidecar 客户端与生命周期管理
 *
 * 架构定位（重要）：sidecar 与 blender-mcp 是【并列】关系，不是串联。
 *
 *   慢回路(秒级):   VCP -> BlenderBridge -> blender-mcp:6090 -> Blender Add-on
 *   快回路(~100ms): Blender GMR Add-on -> sidecar:6091        [不经过 VCP]
 *   管理回路:       VCP -> BlenderBridge -> gmr ex -> sidecar:6091
 *
 * 交互期的拖拽推理必须由 Blender Add-on 直连 6091。理由已写在 README：
 * blender-mcp 上游单并发(BLENDER_BUSY) + 双跳延迟，塞进去必然请求堆积锁死。
 * 本模块只负责“管理”——起停、健康检查、装载模型、以及慢回路的单次试跑。
 */
const http = require('http');
const fs = require('fs');
const path = require('path');
const { spawn } = require('child_process');
const { CONFIG, ensureDirs } = require('./config');
const registry = require('./registry');

const NL = String.fromCharCode(10);
const PID_FILE = () => path.join(CONFIG.root, 'sidecar.pid');
const SIDECAR_LOG = () => path.join(CONFIG.root, 'sidecar.log');

function httpJson(method, urlPath, body, timeoutMs) {
  return new Promise((resolve, reject) => {
    const payload = body === undefined ? null : JSON.stringify(body);
    const headers = { 'Accept': 'application/json' };
    if (payload) {
      headers['Content-Type'] = 'application/json';
      headers['Content-Length'] = Buffer.byteLength(payload);
    }
    const req = http.request({
      host: CONFIG.sidecarHost,
      port: CONFIG.sidecarPort,
      path: urlPath,
      method: method,
      headers: headers,
      timeout: timeoutMs || CONFIG.inferTimeoutMs,
    }, (res) => {
      let raw = '';
      res.setEncoding('utf8');
      res.on('data', (c) => { raw += c; });
      res.on('end', () => {
        let parsed = null;
        try {
          parsed = raw.trim() ? JSON.parse(raw) : {};
        } catch (e) {
          parsed = { raw: raw.slice(0, 500) };
        }
        if (res.statusCode >= 400) {
          const msg = (parsed && parsed.error) ? parsed.error : raw.slice(0, 300);
          return reject(new Error('sidecar HTTP ' + res.statusCode + ': ' + msg));
        }
        resolve(parsed);
      });
    });
    req.on('timeout', () => {
      req.destroy(new Error('sidecar 请求超时 (' + (timeoutMs || CONFIG.inferTimeoutMs) + 'ms)'));
    });
    req.on('error', (err) => {
      if (err.code === 'ECONNREFUSED') {
        return reject(new Error('无法连接推理 sidecar (' + CONFIG.sidecarUrl + ')。'
          + NL + '请先用 sidecar_start 启动，或确认 GMR_SIDECAR_PORT 配置一致。'
          + NL + '注意 sidecar 需要装有 PyTorch 的独立 Python 环境（用 GMR_PYTHON 指定）。'));
      }
      reject(err);
    });
    if (payload) req.write(payload);
    req.end();
  });
}

function isAlive(pid) {
  if (!pid) return false;
  try {
    process.kill(pid, 0);
    return true;
  } catch (e) {
    return e.code === 'EPERM';
  }
}

function readPidFile() {
  const p = PID_FILE();
  if (!fs.existsSync(p)) return null;
  const n = parseInt(fs.readFileSync(p, 'utf8').trim(), 10);
  return Number.isFinite(n) ? n : null;
}

function tail(p, maxBytes) {
  if (!fs.existsSync(p)) return '';
  const size = fs.statSync(p).size;
  const limit = maxBytes || CONFIG.logTailBytes;
  const start = Math.max(0, size - limit);
  const fd = fs.openSync(p, 'r');
  try {
    const len = size - start;
    if (len <= 0) return '';
    const buf = Buffer.alloc(len);
    fs.readSync(fd, buf, 0, len, start);
    return buf.toString('utf8');
  } finally {
    fs.closeSync(fd);
  }
}

async function health() {
  ensureDirs();
  const pid = readPidFile();
  const out = {
    endpoint: CONFIG.sidecarUrl,
    pidFile: PID_FILE(),
    recordedPid: pid,
    processAlive: isAlive(pid),
    logFile: SIDECAR_LOG(),
  };
  try {
    const res = await httpJson('GET', '/health', undefined, CONFIG.healthTimeoutMs);
    out.reachable = true;
    out.sidecar = res;
  } catch (e) {
    out.reachable = false;
    out.error = e.message;
    if (out.processAlive) {
      out.hint = '进程存活但 HTTP 不通：可能仍在加载模型（大 checkpoint 需要时间），或端口不一致。查看日志尾部。';
    }
    out.logTail = tail(SIDECAR_LOG(), 1500);
  }
  return out;
}

async function start(input) {
  ensureDirs();
  const existing = readPidFile();
  if (isAlive(existing)) {
    return {
      started: false,
      pid: existing,
      note: 'sidecar 已在运行（pid ' + existing + '）。如需重启请先 sidecar_stop。',
    };
  }

  const script = path.resolve(String((input && input.script) || path.join(CONFIG.pythonDir, 'sidecar.py')));
  if (!fs.existsSync(script)) {
    throw new Error('sidecar 脚本不存在: ' + script);
  }

  const args = [script, '--host', CONFIG.sidecarHost, '--port', String(CONFIG.sidecarPort)];

  // 可选：启动时预载某个已注册模型
  if (input && input.model_id) {
    const m = registry.getModel(input.model_id);
    args.push('--model', m.path, '--model-kind', m.kind, '--model-id', m.id);
    if (m.joints) args.push('--joints', String(m.joints));
    if (m.fps) args.push('--fps', String(m.fps));
    if (m.upAxis) args.push('--up-axis', String(m.upAxis));
  }
  if (input && input.device) args.push('--device', String(input.device));

  const logFd = fs.openSync(SIDECAR_LOG(), 'a');
  let child;
  try {
    child = spawn(CONFIG.python, args, {
      cwd: CONFIG.pythonDir,
      detached: true,
      stdio: ['ignore', logFd, logFd],
      env: Object.assign({}, process.env, { PYTHONUNBUFFERED: '1' }),
    });
  } catch (e) {
    fs.closeSync(logFd);
    throw new Error('无法启动 sidecar（检查 GMR_PYTHON="' + CONFIG.python + '" 是否为装有 PyTorch 的解释器）: ' + e.message);
  }
  let spawnError = null;
  child.on('error', function (err) { spawnError = err.message; });
  const pid = child.pid;
  child.unref();
  fs.closeSync(logFd);
  fs.writeFileSync(PID_FILE(), String(pid), 'utf8');

  if (spawnError) {
    throw new Error('spawn sidecar 失败: ' + spawnError);
  }

  return {
    started: true,
    pid: pid,
    endpoint: CONFIG.sidecarUrl,
    logFile: SIDECAR_LOG(),
    note: '已在独立进程启动，插件不等待其就绪。模型加载可能需数十秒，请稍后用 sidecar_status 确认 reachable=true。',
  };
}

function stop() {
  const pid = readPidFile();
  if (!pid) return { stopped: false, note: '无 pid 记录，sidecar 可能未通过本插件启动。' };
  if (!isAlive(pid)) {
    try { fs.unlinkSync(PID_FILE()); } catch (e) { /* 忽略 */ }
    return { stopped: false, pid: pid, note: '进程已不存在，已清理 pid 文件。' };
  }
  try {
    process.kill(pid, 'SIGTERM');
  } catch (e) {
    throw new Error('发送 SIGTERM 失败: ' + e.message);
  }
  try { fs.unlinkSync(PID_FILE()); } catch (e) { /* 忽略 */ }
  return { stopped: true, pid: pid, note: '已发送 SIGTERM。若无响应需手动 kill -9 ' + pid + '。' };
}

async function loadModel(input) {
  const id = String((input && input.model_id) || '').trim();
  if (!id) throw new Error('sidecar_load_model 需要参数 model_id。用 list_models 查看。');
  const m = registry.getModel(id);
  if (!fs.existsSync(m.path)) {
    throw new Error('模型文件缺失: ' + m.path + '（注册项仍在，但文件已被移动或删除）');
  }
  const res = await httpJson('POST', '/load_model', {
    model_id: m.id,
    path: m.path,
    kind: m.kind,
    format: m.format,
    joints: m.joints,
    fps: m.fps,
    window: m.window,
    up_axis: m.upAxis,
    feature_dim: m.featureDim,
    device: (input && input.device) || undefined,
  }, 120000);
  return { loaded: m.id, kind: m.kind, sidecar: res };
}

/*
 * 慢回路的单次生成试跑。用途是验证链路与调参，不是交互回路。
 * 真正的拖拽实时生成由 Blender GMR Add-on 直连 6091，不经过这里。
 */
async function infer(input) {
  const body = {};
  if (input.constraints !== undefined) {
    body.constraints = typeof input.constraints === 'string' ? JSON.parse(input.constraints) : input.constraints;
  }
  if (input.window !== undefined) {
    body.window = typeof input.window === 'string' ? JSON.parse(input.window) : input.window;
  }
  if (input.seed !== undefined) body.seed = parseInt(input.seed, 10);
  if (input.steps !== undefined) body.steps = parseInt(input.steps, 10);
  if (input.base_motion !== undefined) body.base_motion = input.base_motion;
  if (input.model_id !== undefined) body.model_id = String(input.model_id);

  if (!body.constraints) {
    throw new Error('infer 需要参数 constraints。格式示例：'
      + NL + '[{"type":"sparse","joint":"foot_L","frame":30,"loc":[0.4,0,0.05]},'
      + NL + ' {"type":"fullbody","frame":0,"pose":[...]}]'
      + NL + '这与 Blender 侧挂在 Empty 上的 gmr_* 自定义属性是同一套契约。');
  }

  const t0 = Date.now();
  const res = await httpJson('POST', '/infer', body, CONFIG.inferTimeoutMs);
  const ms = Date.now() - t0;

  // 动作数据可能很大，不能整份塞进上下文
  const out = {
    latencyMs: ms,
    modelId: res.model_id,
    frames: res.frames,
    joints: res.joints,
    seed: res.seed,
  };
  if (res.output_file) out.outputFile = res.output_file;
  if (res.motion && Array.isArray(res.motion)) {
    out.motionPreview = {
      totalFrames: res.motion.length,
      firstFrame: res.motion[0],
      note: '完整动作数据已省略以免污染上下文。请让 sidecar 写文件并用 output_file 取回。',
    };
  }
  if (ms > 300) {
    out.perfWarning = '单次推理 ' + ms + 'ms。交互式拖拽需 ~100-150ms，超出会失去 rig 手感。'
      + '优化方向：减少采样步数（DDIM/LCM）、缩短窗口、或改用蒸馏模型。';
  }
  return out;
}

module.exports = { health: health, start: start, stop: stop, loadModel: loadModel, infer: infer, tail: tail };

File: Plugin/BlenderBridge/gmr/python/sidecar.py
#!/usr/bin/env python3
"""
GMR 推理 sidecar。

【架构定位】本进程是快回路的服务端，监听 127.0.0.1:6091。

    快回路(~100ms):  Blender GMR Add-on  ->  本进程     [不经过 VCP]
    管理回路(秒级):  VCP -> BlenderBridge -> gmr ex -> 本进程

拖拽推理必须由 Blender Add-on 直连本进程。绝不能走 blender-mcp——
上游单并发(BLENDER_BUSY) + 双跳延迟，交互推理塞进去必然请求堆积锁死。

【为什么用标准库 http.server】
零依赖即可启动。S0 阶段没装 PyTorch 也能跑通全链路：无模型时 /infer 走
三次贝塞尔插值兜底，正好对应论文"用传统插值顶替 ML-Betweener"的第一步。

【接口契约（改动需同步 sidecar.js）】
  GET  /health      -> {status, loaded, device, torch_available, backend, ...}
  POST /load_model  <- {model_id, path, kind, format, joints, fps, window, up_axis, feature_dim, device}
  POST /infer       <- {constraints:[...], window:[a,b], seed, steps, model_id, base_motion}
                    -> {model_id, frames, joints, seed, motion|output_file, backend}
  POST /unload      -> 卸载当前模型

【约束契约】constraints 每项对应 Blender 侧挂在 Empty 上的 gmr_* 自定义属性：
  {"type":"sparse",   "joint":"foot_L", "frame":30, "loc":[x,y,z], "rot":[...]}
  {"type":"fullbody", "frame":0,        "pose":[[...], ...]}
这是三条回路共享的同一套语言。
"""

import argparse
import json
import os
import sys
import threading
import time
import traceback
from datetime import datetime, timezone
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

# 与本文件同目录
import model_config as mc

NL = chr(10)

# sidecar 是常驻进程，状态可留内存（与 Node 侧每次新进程不同）
STATE = {
    "loaded": None,
    "model": None,
    "device": "cpu",
    "started_at": None,
    "infer_count": 0,
    "last_latency_ms": None,
    # 模型配置表与由它解析出的采样入口。启动时加载一次，热重载走 /reload_config。
    "config": None,
    "config_error": None,
    "sampler": None,
}
STATE_LOCK = threading.Lock()


def load_gmr_config(path=None, quiet=False):
    """
    加载模型配置表并解析采样入口。

    配置错误不使进程退出——sidecar 是常驻服务，退出会让 Blender 侧连接直接断掉。
    改为记录 config_error 并降级到贝塞尔兜底，/health 里可见错误原因。
    """
    try:
        cfg = mc.load_config(path, quiet=quiet)
    except ValueError as e:
        with STATE_LOCK:
            STATE["config"] = None
            STATE["config_error"] = str(e)
            STATE["sampler"] = None
        print("[gmr-sidecar] 配置加载失败，降级贝塞尔兜底: " + str(e), file=sys.stderr, flush=True)
        return None

    sampler = None
    err = None
    if cfg:
        try:
            sampler = mc.resolve(cfg, "sampler", "sample")
        except ValueError as e:
            err = str(e)
            print("[gmr-sidecar] 采样入口解析失败，降级贝塞尔兜底: " + str(e), file=sys.stderr, flush=True)

    with STATE_LOCK:
        STATE["config"] = cfg
        STATE["config_error"] = err
        STATE["sampler"] = sampler
    return cfg


def now_iso():
    return datetime.now(timezone.utc).astimezone().isoformat()


def torch_available():
    try:
        import torch  # noqa: F401
        return True
    except ImportError:
        return False


def resolve_device(pref):
    if pref and pref != "auto":
        return pref
    try:
        import torch
        return "cuda" if torch.cuda.is_available() else "cpu"
    except ImportError:
        return "cpu"


def load_model_impl(spec):
    """
    按配置表载入模型。不需要改本函数即可接入模型——填 gmr/model_config.json
    的 model.module 与 model.loader 即可。

    loader 的契约是 loader(config, checkpoint_path, device) -> model。

    安全要点：loader 实现里若用 torch.load 读 .pt/.pth，务必 weights_only=True。
    否则 pickle 反序列化等同于执行任意代码——这也是 registry.js 的
    SAFE_FORMATS 只列 safetensors/onnx/npz 的原因。

    未配置 model.loader 时返回 None，/infer 走贝塞尔兜底。
    """
    path = spec.get("path")
    if path and not os.path.exists(path):
        raise FileNotFoundError("模型文件不存在: " + str(path))

    with STATE_LOCK:
        cfg = STATE["config"]

    if not cfg:
        return None

    loader = mc.resolve(cfg, "model", "loader")
    if loader is None:
        print("[gmr-sidecar] 配置未提供 model.loader，无法载入权重，将走贝塞尔兜底。", flush=True)
        return None

    device = mc.resolve_device(cfg, spec.get("device"))
    model = loader(cfg, path, device)
    if model is None:
        raise ValueError("model.loader 返回了 None。请检查其实现是否返回模型实例。")
    print("[gmr-sidecar] 模型已载入: " + str(path) + " device=" + device, flush=True)
    return model


def cubic_bezier_scalar(p0, p1, p2, p3, t):
    """标准三次贝塞尔，用于骨架期的无模型兜底插值。"""
    mt = 1.0 - t
    return (mt * mt * mt * p0
            + 3.0 * mt * mt * t * p1
            + 3.0 * mt * t * t * p2
            + t * t * t * p3)


def bezier_fallback(constraints, frame_start, frame_end):
    """
    无模型兜底：稀疏约束按帧排序，相邻两点间做三次贝塞尔插值。

    注意：控制点取三分位时三次贝塞尔精确退化为线性插值
    B(t) = a + (b - a) * t，故轨迹是分段线性、节点处一阶导不连续。
    这对验证链路的用途足够；若需真正的 C1 连续，需按相邻段斜率
    推导切线（Catmull-Rom，或 Blender F-Curve 的 auto-clamped handle）。

    这不是"生成"只是插值，但足以验证整条链路，
    并对应论文第一阶段"用传统插值顶替 ML-Betweener"的做法。
    """
    pts = []
    for c in constraints:
        frame = c.get("frame")
        loc = c.get("loc")
        if frame is None or not loc:
            continue
        pts.append((int(frame), [float(v) for v in loc]))
    pts.sort(key=lambda x: x[0])

    if not pts:
        raise ValueError("constraints 中没有可用的 (frame, loc) 数据点。")

    if len(pts) == 1:
        single = pts[0][1]
        return [list(single) for _ in range(frame_start, frame_end + 1)]

    motion = []
    for f in range(frame_start, frame_end + 1):
        if f <= pts[0][0]:
            motion.append(list(pts[0][1]))
            continue
        if f >= pts[-1][0]:
            motion.append(list(pts[-1][1]))
            continue
        seg = 0
        for i in range(len(pts) - 1):
            if pts[i][0] <= f <= pts[i + 1][0]:
                seg = i
                break
        f0, p_a = pts[seg]
        f1, p_b = pts[seg + 1]
        span = max(1, f1 - f0)
        t = (f - f0) / float(span)
        vals = []
        for axis in range(min(len(p_a), len(p_b))):
            a = p_a[axis]
            b = p_b[axis]
            c1 = a + (b - a) / 3.0
            c2 = a + 2.0 * (b - a) / 3.0
            vals.append(cubic_bezier_scalar(a, c1, c2, b, t))
        motion.append(vals)
    return motion


def infer_impl(req):
    """
    生成推理。采样入口由 gmr/model_config.json 的 sampler.module + sampler.sample 指定。

    sample 的契约是 sample(model, constraints, window, seed, config) -> list[list[float]]。

    延迟要求：交互式拖拽需 ~100-150ms。原生 DDPM 1000 步绝无可能，
    必须用 DDIM/LCM 少步采样或蒸馏模型。这是"是 rig 还是批处理"的分水岭。
    """
    constraints = req.get("constraints") or []
    if not isinstance(constraints, list) or not constraints:
        raise ValueError("constraints 必须是非空数组。")

    window = req.get("window")
    if isinstance(window, list) and len(window) == 2:
        f_start, f_end = int(window[0]), int(window[1])
    else:
        frames = [int(c["frame"]) for c in constraints if c.get("frame") is not None]
        if not frames:
            raise ValueError("未提供 window，且 constraints 中无 frame 字段，无法确定时间范围。")
        f_start, f_end = min(frames), max(frames)
    if f_end < f_start:
        raise ValueError("window 非法: [" + str(f_start) + ", " + str(f_end) + "]")

    seed = req.get("seed")
    if seed is None:
        seed = 0

    with STATE_LOCK:
        model = STATE["model"]
        loaded = STATE["loaded"]
        cfg = STATE["config"]
        sampler = STATE["sampler"]

    # 允许单次请求覆盖采样步数，便于在拖拽期用更少步数换低延迟
    if cfg and req.get("steps") is not None:
        cfg = json.loads(json.dumps({k: v for k, v in cfg.items() if k != "_meta"}))
        cfg.setdefault("sampler", {})["steps"] = int(req["steps"])

    t0 = time.time()
    if model is not None and sampler is not None:
        motion = sampler(model, constraints, [f_start, f_end], seed, cfg)
        if motion is None:
            raise ValueError("sampler.sample 返回了 None。契约要求返回 list[list[float]]。")
        motion = list(motion)
        backend = "model"
        joints = (loaded or {}).get("joints") or ((cfg or {}).get("skeleton") or {}).get("joints") or 1
        note = None
    else:
        motion = bezier_fallback(constraints, f_start, f_end)
        backend = "bezier_fallback"
        joints = 1
        if model is None and sampler is None:
            reason = "未载入模型且未配置采样入口"
        elif model is None:
            reason = "已配置采样入口但未载入模型（用 sidecar_load_model 装载）"
        else:
            reason = "已载入模型但未配置 sampler.sample"
        note = ("走三次贝塞尔插值兜底（" + reason + "）。"
                "这验证了链路但不是生成。配置 gmr/model_config.json 并载入模型后自动切换。")

    latency = int((time.time() - t0) * 1000)
    with STATE_LOCK:
        STATE["infer_count"] += 1
        STATE["last_latency_ms"] = latency

    out = {
        "model_id": (loaded or {}).get("model_id"),
        "backend": backend,
        "frames": len(motion),
        "joints": joints,
        "seed": seed,
        "window": [f_start, f_end],
        "latency_ms": latency,
        "motion": motion,
        "note": note,
    }

    # 数据量大时写文件而非塞进响应，避免污染 Agent 上下文
    if len(motion) > 200:
        out_dir = os.environ.get("GMR_MODELS_DIR") or os.getcwd()
        out_path = os.path.join(out_dir, "last_infer.json")
        try:
            with open(out_path, "w", encoding="utf-8") as f:
                json.dump({"motion": motion, "window": [f_start, f_end]}, f)
            out["output_file"] = out_path
            out["motion"] = motion[:3]
            out["note"] = (note or "") + " 完整数据已写入 output_file。"
        except Exception as e:
            out["write_warning"] = str(e)

    return out


class Handler(BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.1"

    def log_message(self, fmt, *args):
        # 默认实现每条请求都刷 stderr；拖拽期每秒数十条会淹掉日志
        if os.environ.get("GMR_HTTP_VERBOSE") == "1":
            sys.stderr.write("[gmr-http] " + (fmt % args) + NL)

    def _send(self, code, payload):
        body = json.dumps(payload, ensure_ascii=False).encode("utf-8")
        self.send_response(code)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def _read_json(self):
        length = int(self.headers.get("Content-Length") or 0)
        if length <= 0:
            return {}
        raw = self.rfile.read(length)
        try:
            return json.loads(raw.decode("utf-8"))
        except Exception as e:
            raise ValueError("请求体不是合法 JSON: " + str(e))

    def do_GET(self):
        if self.path.rstrip("/") in ("/health", ""):
            with STATE_LOCK:
                payload = {
                    "status": "ok",
                    "loaded": STATE["loaded"],
                    "device": STATE["device"],
                    "started_at": STATE["started_at"],
                    "infer_count": STATE["infer_count"],
                    "last_latency_ms": STATE["last_latency_ms"],
                    "torch_available": torch_available(),
                    "backend": ("model" if (STATE["model"] is not None and STATE["sampler"] is not None)
                                else "bezier_fallback"),
                    "config_loaded": STATE["config"] is not None,
                    "config_path": ((STATE["config"] or {}).get("_meta") or {}).get("path"),
                    "sampler_ready": STATE["sampler"] is not None,
                    "config_error": STATE["config_error"],
                }
            self._send(200, payload)
        else:
            self._send(404, {"error": "未知路径 " + self.path
                             + "。可用: GET /health, POST /load_model, POST /infer, POST /unload"})

    def do_POST(self):
        route = self.path.rstrip("/")
        try:
            req = self._read_json()

            if route == "/load_model":
                spec = dict(req)
                dev = resolve_device(spec.get("device"))
                model = load_model_impl(spec)
                with STATE_LOCK:
                    STATE["model"] = model
                    STATE["device"] = dev
                    STATE["loaded"] = {
                        "model_id": spec.get("model_id"),
                        "kind": spec.get("kind"),
                        "format": spec.get("format"),
                        "path": spec.get("path"),
                        "joints": spec.get("joints"),
                        "fps": spec.get("fps"),
                        "window": spec.get("window"),
                        "up_axis": spec.get("up_axis"),
                        "feature_dim": spec.get("feature_dim"),
                        "loaded_at": now_iso(),
                    }
                warn = ""
                if model is None:
                    warn = ("已登记元数据但未真正载入权重（配置未启用，或 gmr/model_config.json "
                            "缺 model.loader），/infer 仍走贝塞尔兜底。详见 /health 的 config_error。")
                if not spec.get("up_axis"):
                    warn = warn + " 缺 up_axis：Blender 是 Z-up 而多数模型是 Y-up，缺此项会导致角色躺平。"
                resp = {"loaded": STATE["loaded"], "device": dev}
                if warn.strip():
                    resp["warning"] = warn.strip()
                self._send(200, resp)

            elif route == "/infer":
                self._send(200, infer_impl(req))

            elif route == "/unload":
                with STATE_LOCK:
                    STATE["model"] = None
                    STATE["loaded"] = None
                self._send(200, {"unloaded": True})

            elif route == "/reload_config":
                # 热重载配置表，免去改配置就要重启 sidecar、断开 Blender 连接
                cfg = load_gmr_config(req.get("path"), quiet=True)
                with STATE_LOCK:
                    self_err = STATE["config_error"]
                    ready = STATE["sampler"] is not None
                self._send(200, {
                    "reloaded": True,
                    "config_loaded": cfg is not None,
                    "config_path": ((cfg or {}).get("_meta") or {}).get("path"),
                    "sampler_ready": ready,
                    "config_error": self_err,
                    "note": "模型实例未受影响。若更换了 model.loader，需再调用 sidecar_load_model 重新装载。",
                })

            else:
                self._send(404, {"error": "未知路径 " + self.path})

        except NotImplementedError as e:
            self._send(501, {"error": str(e)})
        except (ValueError, FileNotFoundError) as e:
            self._send(400, {"error": str(e)})
        except Exception as e:
            traceback.print_exc()
            self._send(500, {"error": type(e).__name__ + ": " + str(e)})


def main():
    p = argparse.ArgumentParser(description="GMR 推理 sidecar")
    p.add_argument("--host", default="127.0.0.1")
    p.add_argument("--port", type=int, default=6091)
    p.add_argument("--model", default=None, help="启动时预载的权重路径")
    p.add_argument("--model-id", default=None)
    p.add_argument("--model-kind", default="betweener")
    p.add_argument("--joints", type=int, default=None)
    p.add_argument("--fps", type=int, default=None)
    p.add_argument("--up-axis", default=None)
    p.add_argument("--device", default="auto")
    p.add_argument("--config", default=None, help="模型配置表路径，默认 gmr/model_config.json")
    args = p.parse_args()

    STATE["started_at"] = now_iso()

    # 先加载配置表，再定设备——配置里的 runtime.device 应能参与决策
    cfg = load_gmr_config(args.config)
    STATE["device"] = mc.resolve_device(cfg, args.device)
    if cfg:
        print("[gmr-sidecar] " + mc.describe(cfg).replace(NL, NL + "[gmr-sidecar] "), flush=True)
        with STATE_LOCK:
            ready = STATE["sampler"] is not None
        if not ready:
            print("[gmr-sidecar] 警告：配置已启用但采样入口不可用，/infer 仍走贝塞尔兜底。", flush=True)

    print("[gmr-sidecar] 启动 " + args.host + ":" + str(args.port)
          + " device=" + STATE["device"] + " torch=" + str(torch_available()), flush=True)

    if args.model:
        try:
            spec = {
                "model_id": args.model_id,
                "path": args.model,
                "kind": args.model_kind,
                "joints": args.joints,
                "fps": args.fps,
                "up_axis": args.up_axis,
            }
            STATE["model"] = load_model_impl(spec)
            spec["loaded_at"] = now_iso()
            STATE["loaded"] = spec
            print("[gmr-sidecar] 预载完成: " + str(args.model), flush=True)
        except Exception as e:
            print("[gmr-sidecar] 预载失败（服务继续启动）: " + str(e), file=sys.stderr, flush=True)

    if STATE["model"] is None:
        print("[gmr-sidecar] 当前无模型，/infer 走三次贝塞尔兜底（S0 阶段，可验证链路）。", flush=True)

    srv = ThreadingHTTPServer((args.host, args.port), Handler)
    srv.daemon_threads = True
    try:
        print("[gmr-sidecar] 就绪，等待请求。", flush=True)
        srv.serve_forever()
    except KeyboardInterrupt:
        print("[gmr-sidecar] 收到中断，退出。", flush=True)
    finally:
        srv.server_close()


if __name__ == "__main__":
    main()